src/software/ai/motion_constraint/motion_constraint_visitor.cpp

```diff
@@ -0,0 +1,82 @@
+#include "software/ai/motion_constraint/motion_constraint_visitor.h"
+
+#include "software/ai/hl/stp/tactic/all_tactics.h"
+// We disable clang-format here because it makes these lists borderline unreadable,
+// and certainly way more difficult to edit
+// clang-format off
+void MotionConstraintVisitor::visit(CherryPickTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(ShadowFreekickerTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(GoalieTactic &tactic)
+{
+       *current_allowed_constraints = std::set<MotionConstraint>({
+       MotionConstraint::FRIENDLY_DEFENSE_AREA,
+       MotionConstraint::FRIENDLY_DEFENSE_AREA,
+       MotionConstraint::HALF_METER_AROUND_BALL,
+       MotionConstraint::FRIENDLY_HALF
+       });
+}
+
+void MotionConstraintVisitor::visit(CreaseDefenderTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(ShadowEnemyTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(MoveTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(ChipTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(KickoffChipTactic &tactic)
+{
+       *current_allowed_constraints = std::set<MotionConstraint>({
+       MotionConstraint::CENTER_CIRCLE,
+       MotionConstraint::HALF_METER_AROUND_BALL
+       });
+}
+
+void MotionConstraintVisitor::visit(StopTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(PatrolTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(PenaltyKickTactic &tactic)
+{
+        *current_allowed_constraints = std::set<MotionConstraint>({
+        MotionConstraint::HALF_METER_AROUND_BALL,
+        MotionConstraint::ENEMY_DEFENSE_AREA,
+        MotionConstraint::ENEMY_HALF
+        });
+}
+
+void MotionConstraintVisitor::visit(PenaltySetupTactic &tactic)
+{
+        *current_allowed_constraints = std::set<MotionConstraint>({
+        MotionConstraint::ENEMY_HALF,
+        MotionConstraint::ENEMY_DEFENSE_AREA,
+        MotionConstraint::FRIENDLY_HALF,
+        MotionConstraint::HALF_METER_AROUND_BALL
+        });
+}
+
+void MotionConstraintVisitor::visit(ReceiverTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(ShootGoalTactic &tactic)
+{
+        *current_allowed_constraints = std::set<MotionConstraint>({
+        MotionConstraint::HALF_METER_AROUND_BALL
+        });
+}
+
+void MotionConstraintVisitor::visit(PasserTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(DefenseShadowEnemyTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(MoveTestTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(StopTestTactic &tactic) {}
+
+void MotionConstraintVisitor::visit(GoalieTestTactic &tactic) {}
+
+void MotionConstraintVisitor::set_current_allowed_constraint(std::set<MotionConstraint> &constraints) {
+    current_allowed_constraints = &constraints;
+
```

Comment:
`std::set<MotionConstraint> &constraints` will pass in a reference to the std::set into this function. We then use the & operator to get the address of the reference, which I _think_ is undefined? 

Just curious why we store a ptr to the reference, rather than just taking a ptr as an arg directly into the function?

Also is there any way to avoid passing in a ptr to a set to be populated? It would be totally reasonable to return a set from the visitor as well

Reply:
Hmm I can also just pass in the ptr directly to the function. The reason I was using pointers was that I wanted the Visitor to access the same list when being called from the Builder. If it is better to return a new set rather than deal with the pointers, I can work on that change.